securedrop/tests/test_source.py

```diff
@@ -74,462 +72,489 @@ def _find_codename(self, html):
+            assert resp.status_code == 200
+            assert session['logged_in'] is True
+
+            resp = app.get('/logout', follow_redirects=True)
+            assert 'logged_in' not in session
+            assert 'codename' not in session
+            text = resp.data.decode('utf-8')
+            assert 'Thank you for exiting your session!' in text
 
-    def test_user_must_log_in_for_protected_views(self):
-        with self.client as c:
-            resp = c.get('/lookup', follow_redirects=True)
-            self.assertEqual(resp.status_code, 200)
-            self.assertIn("Enter Codename", resp.data)
+    def test_user_must_log_in_for_protected_views(self, source_app):
+        with source_app.test_client() as app:
+            resp = app.get('/lookup', follow_redirects=True)
+            assert resp.status_code == 200
+            text = resp.data.decode('utf-8')
+            assert "Enter Codename" in text
 
-    def test_login_with_whitespace(self):
+    def test_login_with_whitespace(self, source_app):
         """
         Test that codenames with leading or trailing whitespace still work"""
 
-        with self.client as client:
-            def login_test(codename):
-                resp = client.get('/login')
-                self.assertEqual(resp.status_code, 200)
-                self.assertIn("Enter Codename", resp.data)
-
-                resp = client.post('/login', data=dict(codename=codename),
-                                   follow_redirects=True)
-                self.assertEqual(resp.status_code, 200)
-                self.assertIn("Submit Materials", resp.data)
-                self.assertTrue(session['logged_in'])
-                resp = client.get('/logout', follow_redirects=True)
-
-            codename = new_codename(client, session)
-            login_test(codename + ' ')
-            login_test(' ' + codename + ' ')
-            login_test(' ' + codename)
-
-    def _dummy_submission(self, client):
+        def login_test(app, codename):
+            resp = app.get('/login')
+            assert resp.status_code == 200
+            text = resp.data.decode('utf-8')
+            assert "Enter Codename" in text
+
+            resp = app.post('/login', data=dict(codename=codename),
+                            follow_redirects=True)
+            assert resp.status_code == 200
+            text = resp.data.decode('utf-8')
+            assert "Submit Materials" in text, text
+            assert session['logged_in'] is True
+
+            resp = app.get('/logout', follow_redirects=True)
+
+        with source_app.test_client() as app:
+            codename = new_codename(app, session)
+
+        codenames = [
+            codename + ' ',
+            ' ' + codename + ' ',
+            ' ' + codename,
+        ]
+
+        for codename_ in codenames:
+            with source_app.test_client() as app:
+                login_test(app, codename_)
+
+    @staticmethod
+    def _dummy_submission(app):
```

Comment:
Nit: for clarity removal of the old _dummy_submission is in commit **pytest fixtures: test_csrf_error_page** but should be in commit **pytest fixtures: test_initial_submission_notification** where the static version is introduced.

Reply:
I don't move all the tests at once, so I leave it in the old class because there are tests that required it,